experiment/schedule_measure_workers.py

```diff
@@ -0,0 +1,155 @@
+
+    redis_host = experiment_config['redis_host']
+    experiment_filestore = experiment_config['experiment_filestore']
+    local_experiment = experiment_utils.is_local_experiment()
+    cloud_compute_zone = experiment_config.get('cloud_compute_zone')
+    env = {
+        'REDIS_HOST': redis_host,
+        'EXPERIMENT_FILESTORE': experiment_filestore,
+        'EXPERIMENT': experiment,
+        'LOCAL_EXPERIMENT': local_experiment,
+        'CLOUD_COMPUTE_ZONE': cloud_compute_zone,
+    }
+
+    zone = experiment_config['cloud_compute_zone']
+    instance_template_url = gcloud.create_instance_template(
+        instance_template_name, docker_image, env, project, zone)
+
+    instance_group_name = get_instance_group_name(experiment)
+
+    # GCE will create instances for this group in the format
+    # "m-$experiment-$UNIQUE_ID". Use 'm' is short for "measurer".
+    base_instance_name = 'm-' + experiment
+
+    gce.create_instance_group(instance_group_name, instance_template_url,
+                              base_instance_name, project, zone)
+    queue = queue_utils.initialize_queue(redis_host)
+    return queue
+
+
+def teardown(experiment_config: dict):
+    """Teardown all resources used for running measurer workers."""
+    instance_group_name = get_instance_group_name(
+        experiment_config['experiment'])
+    project = experiment_config['cloud_project']
+    zone = experiment_config['cloud_compute_zone']
+    gce.delete_instance_group(instance_group_name, project, zone)
+    gcloud.delete_instance_template(experiment_config['experiment'])
+
+
+def schedule(experiment_config: dict, queue):
+    """Schedule measurer workers. This cannot be called before
+    initialize_measurers."""
+    logger.info('Scheduling measurer workers.')
+
+    # TODO(metzman): This method doesn't seem to correctly take into account
+    # jobs that are running (the API provided by rq doesn't work intuitively).
+    # That is OK for now since scheduling only happens while nothing is being
+    # measured but this should be fixed.
+    jobs = queue_utils.get_all_jobs(queue)
+    counts = collections.defaultdict(int)
+    for job in jobs:
+        counts[job.get_status(refresh=False)] += 1
+
+    num_instances_needed = counts['queued'] + counts['started']
+    num_instances_needed = min(num_instances_needed, MAX_INSTANCES_PER_GROUP)
+
+    logger.info('Scheduling %d workers.', num_instances_needed)
+    instance_group_name = get_instance_group_name(
+        experiment_config['experiment'])
+    project = experiment_config['cloud_project']
+    zone = experiment_config['cloud_compute_zone']
+    num_instances = gce.get_instance_group_size(instance_group_name, project,
+                                                zone)
+
+    # TODO(metzman): Use autoscaling as it probably can deal with quotas more
+    # easily.
+    if not num_instances_needed:
+        # Can't go below 1 instance per group.
```

Comment:
i remember 0 was allowed for instance group, are you sure about this ?

Reply:
Pretty sure, but last time I checked was a few months ago.
I don't think this is worth investigating now.